File: src/main/java/com/ohgnarly/fileripper/factories/FileServiceFactory.java
package com.ohgnarly.fileripper.factories;

import com.ohgnarly.fileripper.models.FileDefinition;
import com.ohgnarly.fileripper.services.FileService;
import com.ohgnarly.fileripper.services.FlatFileService;
import com.ohgnarly.fileripper.services.XmlFileService;
import org.springframework.stereotype.Component;

@Component
public class FileServiceFactory {
    public FileService createFileService(FileDefinition fileDefinition) {
        switch (fileDefinition.getFileType()) {
            case DELIMITED:
            case FIXED:
                return new FlatFileService(fileDefinition);
            case XML:
                return new XmlFileService(fileDefinition);
            default:
                throw new IllegalArgumentException("Invalid file type provided");
        }
    }
}

File: src/main/kotlin/com/ohgnarly/fileripper/factories/FileServiceFactory.kt
package com.ohgnarly.fileripper.factories

import com.ohgnarly.fileripper.models.FileDefinition
import com.ohgnarly.fileripper.models.FileType
import com.ohgnarly.fileripper.services.FileService
import com.ohgnarly.fileripper.services.FlatFileService
import com.ohgnarly.fileripper.services.XmlFileService


internal fun createFileService(fileDefinition: FileDefinition): FileService {
    return when (fileDefinition.fileType) {
        FileType.DELIMITED, FileType.FIXED -> FlatFileService(fileDefinition)
        FileType.XML -> XmlFileService(fileDefinition)
        else -> throw IllegalArgumentException("Invalid file type provided")
    }
}